Refactor and fix clipboard export, refs #13395

djjuhasz opened the pull request:
- Refactor clipboard export code to reduce duplication of code
- Fix coding  standard issues such as long lines, and trailing whitespace
- Add conditional to avoid trying to export external digital objects, and add notice to log
- Fix bug that always set the job name to "XML export" even for CSV exports
- Add more comments

djjuhasz committed:
Refactor and fix clipboard export, refs #13395

- Refactor clipboard export code to reduce duplication of code
- Copy related digital objects when exporting resource metadata to avoid a
  second search to export related digital objects separately
- Add ES search term to filter out all drafts from search results, rather than
  checking publication status and excluding draft records individually
- Add QubitDigitalObject::isLocalFile() method to test if the related digital
  file is hosted by AtoM
- Don't try to export external digital objects from the clipboard
- Fix coding standard issues such as long lines, and trailing whitespace
- Update and supplement code comments

jraddaoui commented on apps/qubit/modules/clipboard/actions/exportAction.class.php:
```diff
@@ -3,18 +3,18 @@
- * You should have received a copy of the GNU General Public License
- * along with Access to Memory (AtoM).  If not, see <http://www.gnu.org/licenses/>.
+ * You should have received a copy of the GNU General Public License along with
+ * Access to Memory (AtoM).  If not, see <http://www.gnu.org/licenses/>.
```
This is going to be needed in **a lot** of files. We still need to decide the standard we want to follow (re: #13317) but both, PSR-2 and PSR-12, state that _"There MUST NOT be a hard limit on line length; the soft limit MUST be 120 characters; lines SHOULD be 80 characters or less"_.

https://www.php-fig.org/psr/psr-12/#23-lines

So maybe we should not worry that much about it, specially in here.

djjuhasz replied:
Yes, you have a good point.  The thing that galls me about the current statement is that it takes just as many lines when it's wrapped at 80 characters. 😜 

I find that PSR statement on line lengths to be so confusing and ambiguous - but I agree that it allows for the current statement as is. 😕

djjuhasz replied:
Reverted

jraddaoui commented on apps/qubit/modules/clipboard/actions/exportAction.class.php:
```diff
@@ -432,30 +433,20 @@ protected function addField($name)
+            );
           }
+
+          $this->helpMessages[] = __(
+            'Digital objects with restricted access or copyright will not'
+            . ' be exported.'
+          );
```
This should not appear for repositories.

djjuhasz replied:
It doesn't, the `$this->digitalObjectsAvailable` flag above is false for the repository type.

jraddaoui commented on lib/job/arActorExportJob.class.php:
```diff
@@ -0,0 +1,91 @@
+<?php
+
+/*
+ * This file is part of the Access to Memory (AtoM) software.
+ *
+ * Access to Memory (AtoM) is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Access to Memory (AtoM) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Access to Memory (AtoM).  If not, see <http://www.gnu.org/licenses/>.
+ */
```
For example ... :stuck_out_tongue:

If you really want to change this, maybe do a search and replace over the entire AtoM folder in a different commit.

jraddaoui commented on lib/job/arActorExportJob.class.php:
```diff
@@ -0,0 +1,91 @@
+    foreach (
+      arElasticSearchPluginUtil::getScrolledSearchResultIdentifiers($search)
+      as $id
+    ) {
```
New line for the bracket. This is one case I'd prefer in one line even if it's higher than 80 chars.

djjuhasz replied:
Fixed

jraddaoui commented on lib/job/arExportJob.class.php:
```diff
@@ -229,126 +241,104 @@ protected function addDigitalObjects(&$zip, &$errors)
+    if (
+      !isset($this->params['includeDigitalObjects'])
+      || !$this->params['includeDigitalObjects']
+    ) {
```
New line for bracket.

djjuhasz replied:
I like the `) {` convention (obviously 😜), but I think you are right that putting open bracket on a new line is canonical. 😢

djjuhasz replied:
Fixed

jraddaoui commented on lib/job/arExportJob.class.php:
```diff
@@ -229,126 +241,104 @@ protected function addDigitalObjects(&$zip, &$errors)
+          && !$digitalObject->hasConditionalCopyright()
+        )
+      )
+    ) {
```
New line for bracket.

jraddaoui commented on lib/job/arExportJob.class.php:
```diff
@@ -229,126 +241,104 @@ protected function addDigitalObjects(&$zip, &$errors)
+        )
+      )
+    ) {
+      // export is allowed
```
Uppercase first.

djjuhasz replied:
Okay, got this in a *third* commit 🤦

jraddaoui commented on lib/job/arInformationObjectCsvExportJob.class.php:
```diff
@@ -23,232 +23,82 @@
+    // Export descendants if option was selected
+    if (!$this->params['current-level-only'])
+    {
+      foreach ($resource->getDescendantsForExport($options) as $item) {
```
New line for bracket.

djjuhasz replied:
Fixed

jraddaoui commented on lib/job/arInformationObjectExportJob.class.php:
```diff
@@ -0,0 +1,172 @@
+    if (
+      QubitSetting::getByNameAndScope('informationobject', 'default_template')
+      == 'rad'
+    ) {
```
New line for bracket. And another one I'd prefer in a single line.

djjuhasz replied:
Done. 😄

jraddaoui commented on lib/job/arInformationObjectExportJob.class.php:
```diff
@@ -0,0 +1,172 @@
+    foreach (
+      arElasticSearchPluginUtil::getScrolledSearchResultIdentifiers($search)
+      as $id
+    ) {
```
New line for bracket. And the same.

djjuhasz replied:
Done.

jraddaoui commented on lib/job/arExportJob.class.php:
```diff
@@ -255,18 +258,34 @@ protected function addDigitalObject($resource, $tempDir)
     if (
       empty($digitalObject)
       || !$this->allowDigitalObjectExport($resource, $digitalObject)
     ) {
```
Not sure if I missed this one.

djjuhasz replied:
Do you mean it doesn't check if it's an external URI like the comment indicates, or the `) {`?

jraddaoui replied:
Yeah, the bracket.

djjuhasz replied:
Fixed

jraddaoui commented on lib/job/arExportJob.class.php:
```diff
@@ -255,18 +258,34 @@ protected function addDigitalObject($resource, $tempDir)
     }
 
+    // Don't try to export an external digital object
+    if (QubitTerm::EXTERNAL_URI_ID == $digitalObject->usageId)
```
I think we may want to check OFFLINE_ID too (I think it comes only from the metadata DIP upload from AM).

djjuhasz replied:
Oops, missed this one in my commit :(

djjuhasz replied:
Addressed in commit 9a5ba5e3a

jraddaoui reviewed (changes requested):
Looks great @djjuhasz, thanks!

Only a couple of important comments, all the other cosmetics.

djjuhasz reviewed (commented):
Feedback addressed :)

jraddaoui reviewed (approved):
Thanks @djjuhasz!

Just a few not addressed comments, the only one that may be important is to check the OFFLINE_ID usage type.

djjuhasz commented:
> Just a few not addressed comments, the only one that may be important is to check the OFFLINE_ID usage type.

Sorry, missed this the first time I went through your comments. :(  I've addressed it now.

jraddaoui reviewed (approved):
:+1:
